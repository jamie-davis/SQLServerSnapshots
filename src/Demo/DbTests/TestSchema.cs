using System;
using System.Collections.Generic;
using System.Text;
using DbTests.TestConfiguration;
using SQLServerSnapshots;
using TestConsoleLib;
using TestConsoleLib.Testing;
using Xunit;

namespace DbTests
{
    public class TestSchema
    {
        [Fact]
        public void SchemaConfigurationIsConsistent()
        {
            //Arrange
            var collection = new SqlSnapshotCollection(Config.ConnectionString);
            
            //Act
            collection.ConfigureSchema("Chess"); //this loads table definitions direct from the database schema. In theory no further config should be required

            //Assert
            var output = new Output();
            output.WrapLine("Default schema");
            output.WriteLine();
            collection.GetSchemaReport(output, true);            
            output.Report.Verify();
        }

        [Fact]
        public void SchemaConfigurationIsRefined()
        {
            //Arrange/Act
            var collection = SchemaObjectMother.MakeCollection();

            //Assert
            var output = new Output();
            output.WrapLine("Refined schema");
            output.WriteLine();
            collection.GetSchemaReport(output, true);            
            output.Report.Verify();
        }
    }
}
