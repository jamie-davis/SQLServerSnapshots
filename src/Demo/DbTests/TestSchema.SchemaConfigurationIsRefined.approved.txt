Refined schema                                                                                 
                                                                                               
Table Definitions                                                                              
-----------------                                                                              
                       Exclude                                                                 
                       From       Include In                                                   
Table Name             Comparison Comparison Defined By Types                                  
---------------------- ---------- ---------- ----------------                                  
[Chess].[AuditPlayer]  False      False      AuditPlayerTable                                  
                                                                                               
                   Is      Is                                                                            Referenced                      
                   Primary Compare Is            Is          Is       Is       Is       Referenced Table Property   Sort      Sort  Date 
    Name           Key     Key     Unpredictable Predictable Required Included Excluded Name             Name       Direction Index Type 
    -------------- ------- ------- ------------- ----------- -------- -------- -------- ---------------- ---------- --------- ----- -----
    PlayerAuditId  True    False   True          False       False    False    False                                                     
    PlayerId       False   False   False         False       False    False    False    [Chess].[Player] PlayerId                        
    AuditStartDate False   False   True          False       False    False    False                                                Local
    AuditEndDate   False   False   True          False       False    False    False                                                Local
                                                                                               
                       Exclude                                                                 
                       From       Include In                                                   
Table Name             Comparison Comparison Defined By Types                                  
---------------------- ---------- ---------- ----------------                                  
[Chess].[AuditRating]  False      False      AuditRatingTable                                  
                                                                                               
                   Is      Is                                                                            Referenced                      
                   Primary Compare Is            Is          Is       Is       Is       Referenced Table Property   Sort      Sort  Date 
    Name           Key     Key     Unpredictable Predictable Required Included Excluded Name             Name       Direction Index Type 
    -------------- ------- ------- ------------- ----------- -------- -------- -------- ---------------- ---------- --------- ----- -----
    AuditRatingId  True    False   True          False       False    False    False                                                     
    PlayerId       False   False   False         False       False    False    False    [Chess].[Player] PlayerId                        
    RatingId       False   False   False         False       False    False    False    [Chess].[Rating] RatingId                        
    AuditStartDate False   False   True          False       False    False    False                                                Local
    AuditEndDate   False   False   True          False       False    False    False                                                Local
    RatingDate     False   False   True          False       False    False    False                                                Local
                                                                                               
                       Exclude                                                                 
                       From       Include In                                                   
Table Name             Comparison Comparison Defined By Types                                  
---------------------- ---------- ---------- ----------------                                  
[Chess].[Player]       False      False      PlayerTable                                       
                                                                                               
             Is      Is                                                                      Referenced                     
             Primary Compare Is            Is          Is       Is       Is       Referenced Property   Sort      Sort  Date
    Name     Key     Key     Unpredictable Predictable Required Included Excluded Table Name Name       Direction Index Type
    -------- ------- ------- ------------- ----------- -------- -------- -------- ---------- ---------- --------- ----- ----
    PlayerId True    False   True          False       False    False    False                                              
    Name     False   False   False         False       False    False    False                          Ascending           
                                                                                               
                       Exclude                                                                 
                       From       Include In                                                   
Table Name             Comparison Comparison Defined By Types                                  
---------------------- ---------- ---------- ----------------                                  
[Chess].[Rating]       False      False      RatingTable                                       
                                                                                               
                   Is      Is                                                                                                                     
                   Primary Compare Is            Is          Is       Is       Is                              Referenced     Sort      Sort  Date
    Name           Key     Key     Unpredictable Predictable Required Included Excluded Referenced Table Name  Property Name  Direction Index Type
    -------------- ------- ------- ------------- ----------- -------- -------- -------- ---------------------- -------------- --------- ----- ----
    RatingId       True    False   True          False       False    False    False                                                              
    PlayerId       False   False   False         False       False    False    False    [Chess].[Player]       PlayerId                           
    RatingSourceId False   False   False         False       False    False    False    [Chess].[RatingSource] RatingSourceId Ascending 1         
    RatingDate     False   False   True          False       False    False    False                                          Ascending 0         
                                                                                               
                       Exclude                                                                 
                       From       Include In                                                   
Table Name             Comparison Comparison Defined By Types                                  
---------------------- ---------- ---------- ----------------                                  
[Chess].[RatingSource] False      False                                                        
                                                                                               
                   Is      Is                                                                      Referenced                     
                   Primary Compare Is            Is          Is       Is       Is       Referenced Property   Sort      Sort  Date
    Name           Key     Key     Unpredictable Predictable Required Included Excluded Table Name Name       Direction Index Type
    -------------- ------- ------- ------------- ----------- -------- -------- -------- ---------- ---------- --------- ----- ----
    RatingSourceId True    False   True          False       False    False    False                                              
                                                                                               
                                                                                               